src/org/zaproxy/zap/extension/ascanrulesBeta/InsecureHTTPMethod.java

```diff
@@ -1,455 +1,702 @@
+                    bytesRead = is.read(buffer);
+                }
+                String response = new String(bos.toByteArray());
+                if (log.isDebugEnabled())
+                    log.debug("Response is " + totalBytesRead + " bytes: \n" + response);
+
+                Matcher m = thirdPartyContentPattern.matcher(response);
+                if (m.matches()) {
+                    log.debug("Response matches expected third party pattern!");
+                    is.close();
+                    os.close();
+                    bos.close();
+                    socket.close();
+                    return true;
+                } else {
+                    log.debug("Response does *not* match expected third party pattern");
+                }
+                is.close();
+                os.close();
+                bos.close();
+                socket.close();
+                return false;
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(
+                            "Could not establish a socket connection to a third party using the CONNECT HTTP method: NULL socket returned, or non-200 response");
+                    log.debug("The status line returned: " + statusLine);
+                }
+                return false;
+            }
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Could not establish a socket connection to a third party using the CONNECT HTTP method", e);
+            }
+        }
+        return false;
+    }
+
+    private final MessageAndEvidence testHttpMethod(String httpMethod) {
+        //httpMethod = httpMethod.toLowerCase().trim();
+
+        //TODO:chose a body that exploits a vulnerability
+        String randomKey = RandomStringUtils.random(15, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+        String randomValue = RandomStringUtils.random(15, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+
+        String requestBody = '"'+randomKey+"\":\""+randomValue+'"'; //"randomKey":"randomValue"
+        if (httpMethod.equals(HttpRequestHeader.TRACK) || httpMethod.equals(HttpRequestHeader.TRACE) || httpMethod.equals(HttpRequestHeader.CONNECT)) {
+            /* We have separate methods for them */
+            return null;
+        }
+        final HttpRequestHeader traceRequestHeader = this.getBaseMsg().getRequestHeader();
+
+
+        traceRequestHeader.setMethod(httpMethod);
+        traceRequestHeader.setVersion(HttpRequestHeader.HTTP11);
+
+        final HttpMessage tracemsg = getNewMsg();
+        tracemsg.setRequestHeader(traceRequestHeader);
+        if(httpMethod.equals(HttpRequestHeader.PUT)||httpMethod.equals(HttpRequestHeader.PATCH)) {
+            tracemsg.setRequestBody(requestBody);
+        }
+        try {
+            // do not follow redirects. That might ruin our day.
+            sendAndReceive(tracemsg, false);

```

Comment:
This could be confusing (maintenance wise), why is the message etc called trace but potentially used for other methods/verbs?